toolchain/check/check.cpp

```diff
@@ -74,46 +77,169 @@ struct UnitInfo {
+                               name_id, SemIR::NameScopeId::Invalid};
+  auto id = context.AddInst(inst);
+  inst.name_scope_id = context.name_scopes().Add(id, enclosing_scope_id);
+  context.insts().Set(id, inst);
+  context.name_scopes().AddEntry(enclosing_scope_id, name_id, id);
+  return {id, inst.name_scope_id};
+}
+
+static auto CacheCopiedNamespace(
+    llvm::DenseMap<SemIR::NameScopeId, SemIR::NameScopeId>& copied_namespaces,
+    SemIR::NameScopeId import_scope_id, SemIR::NameScopeId to_scope_id)
+    -> void {
+  auto [it, success] = copied_namespaces.insert({import_scope_id, to_scope_id});
+  CARBON_CHECK(success || it->second == to_scope_id)
+      << "Copy result for namespace changed from " << import_scope_id << " to "
+      << to_scope_id;
+}
+
+// Copies enclosing name scopes from the import IR. Handles the parent
+// traversal. Returns the NameScope corresponding to the copied
+// import_enclosing_scope_id.
+static auto CopyEnclosingNameScopeFromImportIR(
+    Context& context, SemIR::TypeId namespace_type_id,
+    const SemIR::File& import_sem_ir,
+    SemIR::NameScopeId import_enclosing_scope_id,
+    llvm::DenseMap<SemIR::NameScopeId, SemIR::NameScopeId>& copied_namespaces)
+    -> SemIR::NameScopeId {
+  // Package-level names don't need work.
+  if (import_enclosing_scope_id == SemIR::NameScopeId::Package) {
+    return import_enclosing_scope_id;
   }
+
+  // The scope to add namespaces to. Note this may change while looking at
+  // enclosing scopes, if we encounter a namespace that's already added.
+  auto scope_cursor = SemIR::NameScopeId::Package;
+
+  // Build a stack of enclosing namespace names, with innermost first.
+  llvm::SmallVector<std::pair<SemIR::NameScopeId, SemIR::NameId>>
+      new_namespaces;
+  while (import_enclosing_scope_id != SemIR::NameScopeId::Package) {
+    // If the namespace was already copied, reuse the results.
+    if (auto it = copied_namespaces.find(import_enclosing_scope_id);
+        it != copied_namespaces.end()) {
+      // We inject names at the provided scope, and don't need to keep
+      // traversing parents.
+      scope_cursor = it->second;
+      break;
+    }
+
+    // The namespace hasn't been copied yet, so add it to our list.
+    auto scope = import_sem_ir.name_scopes().Get(import_enclosing_scope_id);
+    auto scope_inst =
+        import_sem_ir.insts().GetAs<SemIR::Namespace>(scope.inst_id);
+    new_namespaces.push_back({scope_inst.name_scope_id, scope_inst.name_id});
+    import_enclosing_scope_id = scope.enclosing_scope_id;
+  }
+
+  // Add enclosing namespace names, starting with the outermost.
+  for (auto import_namespace : llvm::reverse(new_namespaces)) {
+    auto name_id =
+        CopyNameFromImportIR(context, import_sem_ir, import_namespace.second);
+    auto scope = context.name_scopes().Get(scope_cursor);
+    auto [it, success] = scope.names.insert({name_id, SemIR::InstId::Invalid});
```

Comment:
Note that `scope` here is a *copy* of the `NameScope`, so this will be inserting into a copy of the scope's `names` list, not into the list actually stored on the `NameScope`. It took me a while to figure out how this is giving the right results anyway. It turns out that we insert this namespace name twice -- once here, into the copy of the scope list, and again in `AddNamespace` below, into the actual scope.

Aside from not making the `scope` copy, I think we should refactor these insertions so we only do the lookup of `name_id` in `scope_cursor` once.